hex2bytes for Vector{UInt8}

sambitdash opened the pull request:
`function hex2bytes(d::Vector{UInt8}, s::Vector{UInt8}, nInBytes::Int=length(s)) -> Int`

1. The loop is not optimized for SIMD as the branching in the internal loops is too high
2. Computation is carried out on the word boundary. 

Benchmark numbers attached:

```
julia> @benchmark f_hsS_NOSIMD()   <- hex2bytes(::AbstractString)
BenchmarkTools.Trial: 
  memory estimate:  5.00 MiB
  allocs estimate:  2
  --------------
  minimum time:     56.887 ms (0.00% GC)
  median time:      58.451 ms (0.00% GC)
  mean time:        58.693 ms (0.16% GC)
  maximum time:     65.453 ms (0.00% GC)
  --------------
  samples:          86
  evals/sample:     1

julia> 

julia> @benchmark f_hsB_NOSIMD()
BenchmarkTools.Trial: 
  memory estimate:  16 bytes
  allocs estimate:  1
  --------------
  minimum time:     43.811 ms (0.00% GC)
  median time:      44.620 ms (0.00% GC)
  mean time:        44.876 ms (0.00% GC)
  maximum time:     50.646 ms (0.00% GC)
  --------------
  samples:          112
  evals/sample:     1

julia> @benchmark f_hsB_SIMD()
BenchmarkTools.Trial: 
  memory estimate:  16 bytes
  allocs estimate:  1
  --------------
  minimum time:     44.923 ms (0.00% GC)
  median time:      45.786 ms (0.00% GC)
  mean time:        46.127 ms (0.00% GC)
  maximum time:     53.770 ms (0.00% GC)
  --------------
  samples:          109
  evals/sample:     1

```

Fix #23161

sambitdash committed:
hex2bytes for Vector{UInt8}

sambitdash commented:
**Sanitizing input vs. Rejecting Bad Input**

All input was sanitized instead of rejecting for bad input. 

SIMD performance gains are almost 10% due to this:

```
@inline get_sanitized_number_from_hex(c::UInt) = begin
    const DIGIT_NINE     = UInt('9')
    return (c > DIGIT_NINE) ? ((c & 0x07) + 9) : (c & 0x0F)
end

julia> @benchmark f_hsB_SIMD_SANE()
BenchmarkTools.Trial: 
  memory estimate:  16 bytes
  allocs estimate:  1
  --------------
  minimum time:     37.489 ms (0.00% GC)
  median time:      37.915 ms (0.00% GC)
  mean time:        38.010 ms (0.00% GC)
  maximum time:     42.396 ms (0.00% GC)
  --------------
  samples:          132
  evals/sample:     1

julia> @benchmark f_hsB_NOSIMD_SANE()
BenchmarkTools.Trial: 
  memory estimate:  16 bytes
  allocs estimate:  1
  --------------
  minimum time:     42.419 ms (0.00% GC)
  median time:      42.976 ms (0.00% GC)
  mean time:        43.062 ms (0.00% GC)
  maximum time:     47.542 ms (0.00% GC)
  --------------
  samples:          117
  evals/sample:     1

```

JeffBezanson commented:
Thanks, @sambitdash !

Noting that this fixes #23161

JeffBezanson commented on base/strings/util.jl:
```diff
@@ -464,6 +464,55 @@ function hex2bytes(s::AbstractString)
+copied into the destination array. The size of destination array must be at least half of
+the `nInBytes` parameter.
+"""
+function hex2bytes(d::Vector{UInt8}, s::Vector{UInt8}, nInBytes::Int=length(s))
```
Since this modifies its argument, it should be called `hex2bytes!`. It would be good to also add a 1-argument version that allocates the output for you, `hex2bytes(s::Vector{UInt8})`.

JeffBezanson commented on base/strings/util.jl:
```diff
@@ -464,6 +464,55 @@ function hex2bytes(s::AbstractString)
+    end
+
+    len2 = div(nInBytes, 2)
+    if size(d)[1] <  len2
```
Should use `length(d)` here.

JeffBezanson replied:
Also this code uses `@inbounds` but doesn't check that `nInBytes <= length(s)`.

JeffBezanson commented on base/strings/util.jl:
```diff
@@ -464,6 +464,55 @@ function hex2bytes(s::AbstractString)
+    end
+
+    i = 0
+    j = 1
```
This will return 1 if the input is empty.

JeffBezanson commented:
Please add some tests in test/strings/util.jl.

fredrikekre commented on base/strings/util.jl:
```diff
@@ -464,6 +464,55 @@ function hex2bytes(s::AbstractString)
+end
+
+@inline get_number_from_hex(c::UInt) = begin
+    const DIGIT_ZERO     = UInt('0')
```
Remove all `const`.

fredrikekre commented on base/strings/util.jl:
```diff
@@ -464,6 +464,55 @@ function hex2bytes(s::AbstractString)
+    return j
+end
+
+@inline get_number_from_hex(c::UInt) = begin
```
```julia
@inline function number_from_hex(c::UInt)
    # ...
end
```

fredrikekre commented on base/strings/util.jl:
```diff
@@ -464,6 +464,55 @@ function hex2bytes(s::AbstractString)
 end
 
 """
+    hex2bytes(d::Vector{UInt8}, s::Vector{UInt8}, nInBytes::Int=length(s))
```
`nInBytes` is a bit of weird capitalization. Perhaps use `ninbytes` or perhaps even better just `n`,

fredrikekre commented on base/strings/util.jl:
```diff
@@ -464,6 +464,55 @@ function hex2bytes(s::AbstractString)
+results are populated into a destination array. The function returns the number of bytes
+copied into the destination array. The size of destination array must be at least half of
+the `nInBytes` parameter.
+"""
```
Could use and example here:
````

# Examples
```jldoctest
julia> hex2bytes(...)
end
```
````

stevengj replied:
`hex2bytes` now has an example. 

Typically, when we have both an in-place and out-of-place version of a function, we only give examples in the simpler out-of-place version.

sambitdash committed:
Incorporated review comments.

sambitdash committed:
Added example code.

sambitdash commented:
Added test cases.

StefanKarpinski commented:
I'm not entirely sold on the `n` argument. It's going to be hard to remember whether it refers to the number of bytes decoded or the number of hex digits to consume. In fact, I would assume that it indicated the number of bytes to decode, so the number of input bytes would be `2n` – but it means the opposite. Is that parameter really necessary? Can't the number of digits to decode by implied by the length of the output array? If one wants to write into part of a larger output array, it can be done with a view, e.g. `hex2bytes!(@view(output[a:b]), intput)`.

StefanKarpinski commented:
It also should not return the number of bytes decoded – that's completely deterministic based on the input, unlike I/O functions, where it's a useful return value. Instead, APIs like this typically return  the destination object so that you can easily write code that chains operations.

sambitdash commented:
@StefanKarpinski  while destination buffer makes sense, when you are reading from a file of 2020 bytes and your temporary buffer is 1000 bytes, the last 20 bytes read may need a @view or a mechanism to specify the n variable. So may be it would be hex2bytes(d::AbstractArray{UInt8}, s::AbstractArray{UInt8}). One can use @view for both source and destination. But developers may miss to pick up the @view approach. I will modify the signature accordingly.

fredrikekre commented on base/strings/util.jl:
```diff
@@ -464,11 +464,103 @@ function hex2bytes(s::AbstractString)
+    hex2bytes(s::Vector{UInt8})
+
+Convert the hexadecimal bytes array to its binary representation. Returns an
+`Array{UInt8,1}`, i.e. an array of bytes.
```
Perhaps use `Vector{UInt8}` to match the above signature?

fredrikekre commented on base/strings/util.jl:
```diff
@@ -464,11 +464,103 @@ function hex2bytes(s::AbstractString)
 """
+    hex2bytes(s::Vector{UInt8})
+
+Convert the hexadecimal bytes array to its binary representation. Returns an
```
`returns` -> `return`

fredrikekre commented on base/strings/util.jl:
```diff
@@ -464,11 +464,103 @@ function hex2bytes(s::AbstractString)
+`Array{UInt8,1}`, i.e. an array of bytes.
+"""
+@inline function hex2bytes(s::Vector{UInt8})
+    d = Vector{UInt8}(div(length(s), 2))
```
Probably error here if the length is not even?

sambitdash replied:
Not needed as the underlying method will throw exception. Per @StefanKarpinski  all the interfaces are changed to AbstractVector. Many of comments may not be relevant in that case.

fredrikekre commented on base/strings/util.jl:
````diff
@@ -464,11 +464,103 @@ function hex2bytes(s::AbstractString)
+of the `n` parameter.
+
+# Examples
+```
````
````
```jldoctest
````

fredrikekre commented on base/strings/util.jl:
````diff
@@ -464,11 +464,103 @@ function hex2bytes(s::AbstractString)
+
+# Examples
+```
+    julia> s = UInt8["01abEF"...]
````
no indentation here and below

fredrikekre commented on base/strings/util.jl:
```diff
@@ -464,11 +464,103 @@ function hex2bytes(s::AbstractString)
 Convert an array of bytes to its hexadecimal representation.
 All characters are in lower-case.
-
+it
```
?

fredrikekre commented on test/strings/util.jl:
```diff
@@ -247,3 +247,24 @@ bin_val = hex2bytes("07bf")
 #non-hex characters
 @test_throws ArgumentError hex2bytes("0123456789abcdefABCDEFGH")
+
+function test_23161()
```
perhaps rewrite as a `@testset`?

sambitdash committed:
Modified to support AbstractArray

sambitdash committed:
Fixed comments and changed to AbstractVector

sambitdash commented:
All tests are added and all the review comments are incorporated.

fredrikekre commented on base/strings/util.jl:
```diff
@@ -464,11 +464,91 @@ function hex2bytes(s::AbstractString)
+ 0x00
+ 0x00
+
+julia> hex2bytes!(d, s)
```
This returns `d` now, not 3, right? The docstring above should be changed in regard to this as well.

sambitdash replied:
Example updated.

fredrikekre replied:
Also
> The docstring above should be changed in regard to this as well.

sambitdash replied:
done.

sambitdash committed:
Example output fixed.

sambitdash committed:
updated the description.

tkelman commented on base/exports.jl:
```diff
@@ -762,6 +762,7 @@ export
     graphemes,
     hex,
     hex2bytes,
+    hex2bytes!,
```
this will need to be added to the stdlib doc index somewhere to show up in the rendered docs https://github.com/JuliaLang/julia/blob/master/CONTRIBUTING.md#adding-a-new-docstring-to-base

sambitdash replied:
done.

sambitdash committed:
Add method to stdlib index

fredrikekre commented on base/strings/util.jl:
```diff
@@ -464,11 +464,87 @@ function hex2bytes(s::AbstractString)
+        n = 0
+        c1, i = next(s, i)
+        done(s, i) && throw(ArgumentError(
+            "string length must be even: length($(repr(s))) == $(length(s))"))
```
Can this check be done on entry? Should probably not print the string, enough with the length.

sambitdash replied:
This check on beginning means you have a potential case of multiple passes of the buffer as there is no clarity in a AbstractVector if the size is pre-computed. And depend on the implementation details. Is there a significant benefit in precheck? Secondly the functionality cannot be achieved without a full scan due to the nature of the alogorithm used. Hence, my personal preference will be computation on failure.

fredrikekre commented on base/strings/util.jl:
```diff
@@ -464,11 +464,87 @@ function hex2bytes(s::AbstractString)
+        n += number_from_hex(c2)
+        d[j+=1] = (n & 0xFF)
+    end
+    resize!(d, j)
```
This won't work with arbitrary `AbstractVector`, ref https://github.com/JuliaLang/julia/pull/23267#issuecomment-322566856

sambitdash replied:
dependency on length removed. The array should not be printed as it an be large. 

-

fredrikekre replied:
```julia
julia> x = Vector{UInt8}(10); resize!(view(x, 1:10), 8)
ERROR: MethodError: no method matching resize!(::SubArray{UInt8,1,Array{UInt8,1},Tuple{UnitRange{Int64}},true}, ::Int64)
Closest candidates are:
  resize!(::Array{T,1} where T, ::Integer) at array.jl:1020
  resize!(::BitArray{1}, ::Integer) at bitarray.jl:836
```

fredrikekre commented on base/strings/util.jl:
```diff
@@ -464,11 +464,87 @@ function hex2bytes(s::AbstractString)
+    return (DIGIT_ZERO <= c <= DIGIT_NINE) ? c - DIGIT_ZERO :
+        (LATIN_UPPER_A <= c <= LATIN_UPPER_F) ? c - LATIN_UPPER_A + 10 :
+        (LATIN_A <= c <= LATIN_F) ? c - LATIN_A + 10 :
+        throw(ArgumentError("Not a hexadecimal number"))
```
perhaps `"$c is not a hexadecimal number"`

sambitdash replied:
done

sambitdash committed:
length should not be used with AbstractArray()

in exception handling as it can raise exception.

sambitdash committed:
removed resize as it may not work for subarrays.

sambitdash committed:
Added subarray test cases.

KristofferC commented on base/strings/util.jl:
```diff
@@ -464,11 +464,89 @@ function hex2bytes(s::AbstractString)
+ 0x45
+ 0x46
+
+julia> d =zeros(UInt8, 3)
```
missing space

sambitdash replied:
fixed in last check-in.

KristofferC commented on base/strings/util.jl:
```diff
@@ -464,11 +464,89 @@ function hex2bytes(s::AbstractString)
+    LATIN_A        = UInt('a')
+    LATIN_F        = UInt('f')
+
+    return (DIGIT_ZERO <= c <= DIGIT_NINE) ? c - DIGIT_ZERO :
```
3 nested tertiaries is a bit hard to reason about. Perhaps something like

```jl
DIGIT_ZERO    <= c <= DIGIT_NINE    && return c - DIGIT_ZERO
LATIN_UPPER_A <= c <= LATIN_UPPER_F && return c - LATIN_UPPER_A + 10
LATIN_A       <= c <= LATIN_F       && return c - LATIN_A + 10
throw(ArgumentError("not a hexadecimal number: '$(Char(c))'"))
```

sambitdash replied:
This is a matter of personal preference and choice. A few lines above in hex2bytes(AbstractVector) has 3 nested  tertiaries. Hence ignoring this comment.

sambitdash replied:
This is a matter of personal preference and choice. A few lines above in hex2bytes(AbstractVector) has 3 nested  tertiaries. Hence ignoring this comment.

KristofferC replied:
Yeah, we can fix it up at some other point, doesn't need to block this PR.

sambitdash committed:
typo fixed

fredrikekre commented:
Oh, I just saw that there are a method of `hex2bytes(::AbstractString)` just above, it should be possible to collapse these, now it is very much code duplication.

sambitdash commented:
Some code duplications are naturally desired. If you plan to convert string to byte array or vice versa before using one method over the other then you are essentially going to add substantial performance overheads. A typical usage should thought in multi-megabyte byte array as input and output. Anyway there is already a penalty additional to supporting subarrays. 

When I proposed the changes each change had a benchmark carried out. If you plan to make additional changes to the code would suggest you conduct necessary benchmark and proceed from here. I will not be in a position to devote further efforts on this.

fredrikekre commented:
Sure, can be fixed later.

> Anyway there is already a penalty additional to supporting subarrays.

How so?

sambitdash commented:
Simple array indexing is faster than iterations used. You can run benchmark on that.

sambitdash commented:
Ok here is a benchmark, I just carried out with iterators vs. full array. There is a minor 5% penalty on using the iterator over array indexing. But guess the flexibility is far superior to support SubArrays. Hence, it's a good thing that we are supporting `AbstractVector` as interface. The input array is a random string of 10mb as input. 

```
julia> @benchmark f_hsB_SubArray_noview()
BenchmarkTools.Trial: 
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     43.614 ms (0.00% GC)
  median time:      44.429 ms (0.00% GC)
  mean time:        44.783 ms (0.00% GC)
  maximum time:     54.849 ms (0.00% GC)
  --------------
  samples:          112
  evals/sample:     1

julia> @benchmark f_hsB_NOSIMD()
BenchmarkTools.Trial: 
  memory estimate:  16 bytes
  allocs estimate:  1
  --------------
  minimum time:     41.739 ms (0.00% GC)
  median time:      42.399 ms (0.00% GC)
  mean time:        42.464 ms (0.00% GC)
  maximum time:     47.865 ms (0.00% GC)
  --------------
  samples:          118
  evals/sample:     1
```

sambitdash commented:
```
const mb_10 = (10 << 20)

arr=UInt8[rand(['0','1','2','3','4','5','6','7','8','9','0','a','b','c','d','e','f'])  for x = 1:mb_10]
arr1=zeros(UInt8, (10<<19))
```
Input and output buffer composition for benchmark.

stevengj commented on base/strings/util.jl:
```diff
@@ -464,11 +464,89 @@ function hex2bytes(s::AbstractString)
+    return d
+end
+
+@inline function number_from_hex(c::UInt)
```
Why is this `UInt` and not `UInt8`?

stevengj replied:
And it seems like this function should return a `UInt8` as well.

sambitdash replied:
Non-word boundary computation can be 10 times slower or that's what was observed.

stevengj commented on base/strings/util.jl:
```diff
@@ -464,11 +464,89 @@ function hex2bytes(s::AbstractString)
+    if isodd(len)
+        throw(ArgumentError("source vector length must be even"))
+    end
+    d = zeros(UInt8, div(len, 2))
```
Why initialize to zero?  Just use `Vector{UInt8}(len >> 1)` to allocate an uninitialized array.

stevengj commented on base/strings/util.jl:
```diff
@@ -464,11 +464,89 @@ function hex2bytes(s::AbstractString)
+    # byte boundary. Boundary computation can be almost 10 times slower
+    n::UInt = 0
+    c1::UInt = 0
+    c2::UInt = 0
```
Why do you need explicit type declarations here?  Why do you need to initialize `c1` and `c2` at all?  The compiler will figure out that `next` returns a `UInt8`.

sambitdash replied:
Because we want to force computation to word boundary.

stevengj commented on base/strings/util.jl:
```diff
@@ -464,11 +464,89 @@ function hex2bytes(s::AbstractString)
+    c1::UInt = 0
+    c2::UInt = 0
+    while !done(s, i)
+        n = 0
```
Why initialize `n` to zero?

stevengj commented on base/strings/util.jl:
```diff
@@ -464,11 +464,89 @@ function hex2bytes(s::AbstractString)
+        n = number_from_hex(c1)
+        n <<= 4
+        n += number_from_hex(c2)
+        d[j+=1] = (n & 0xFF)
```
Seems like it would be clearer to just do:
```
d[j += 1] = number_from_hex(c1) << 4 + number_from_hex(c2)
```
I don't see why the `& 0xFF` mask is needed, since `number_from_hex` should presumably always return a 4-bit quantity.

sambitdash replied:
The whole purpose of introducing UInt8 method was not to deteriorate performance significantly over the hex2bytes(s::AbstractString) method. And in the end it beats its performance by 10-15% at least. I guess before doing any changes to the code in terms of style and coding practices performance benchmarking be carried out. At this point I will not be able to invest further effort on this. But some of these typically compilers automatically identify and optimize but did not seem to be byte boundary computation was one of them. 

The ideal code for this method should be written by taking 4 bytes at a time (sanitize the input rather than error check the input) and optimize the code for SIMD as this is supposed to be a very low level function for IO operations.

Hence, unless ignoring this comment unless there is a supporting benchmark data.

sambitdash replied:
@stevengj ideally your code and my code should generate the same code with most modern compilers. With the code I have written `n` can easily be spotted as a register on the CPU. It's an old practice may be irrelevant in today's compilers.  I have not checked the generated code so used a conservative style of coding. 

(n & 0xff) was a guidance to compiler not to create an `InexactError` exception block. Not sure if Julia actually honor that.

sambitdash replied:
@stevengj ideally your code and my code should generate the same code with most modern compilers. With the code I have written `n` can easily be spotted as a register on the CPU. It's an old practice may be irrelevant in today's compilers.  I have not checked the generated code so used a conservative style of coding. 

(n & 0xff) was a guidance to compiler not to create an `InexactError` exception block. Not sure if Julia actually honor that.

stevengj commented on base/strings/util.jl:
```diff
@@ -464,11 +464,89 @@ function hex2bytes(s::AbstractString)
+    while !done(s, i)
+        n = 0
+        c1, i = next(s, i)
+        done(s, i) && throw(ArgumentError("source vector length must be even"))
```
This check should not occur inside the loop.

In particular, there should probably be a check
```
if 2length(d) != length(s)
    isodd(length(s)) && throw(ArgumentError("input hex array must have even length"))
    throw(ArgumentError("output array must be half length of input array")) 
end
```
And then the `isodd` check can be removed from `hex2bytes`.

sambitdash replied:
For a badly created input of a view which has included alternate data elements considered as view design technically there may be a 2-pass on length calculation. Particularly, the same reason why hex2bytes(s::AbstractString) does not have this check as well. And the intended behavior cannot be achieved without a full pass of the array successfully. 

As commented the high performance need is with the two parameter method. Hence,  ignoring this comment.

stevengj commented on base/strings/util.jl:
```diff
@@ -464,11 +464,89 @@ function hex2bytes(s::AbstractString)
+    LATIN_UPPER_A  = UInt('A')
+    LATIN_UPPER_F  = UInt('F')
+    LATIN_A        = UInt('a')
+    LATIN_F        = UInt('f')
```
I don't think defining these constants adds much to the clarity (on the contrary, it makes the code longer and hence harder to read); it would be just as compact and readable if you used `UInt8('0')` etcetera directly in the code.

sambitdash replied:
This is again an issue of style and personal choice. Some people find balancing parenthesis an unnecessary overhead in code.

stevengj commented on base/strings/util.jl:
````diff
@@ -464,11 +464,89 @@ function hex2bytes(s::AbstractString)
+
+# Examples
+```jldoctest
+julia> s = UInt8["01abEF"...]
````
`b"01abEF"` is the simplest way to do this.

The slightly more verbose way would be `Vector{UInt8}("01abEF")`.   The manual should not recommend using splatting.

stevengj commented on base/strings/util.jl:
```diff
@@ -464,11 +464,89 @@ function hex2bytes(s::AbstractString)
+    hex2bytes(s::AbstractVector{UInt8})
+
+Convert the hexadecimal bytes array to its binary representation. Returns
+`Vector{UInt8}`, i.e. a vector of bytes.
```
This is pretty unclear on what a "hexadecimal bytes vector" and "its binary representation" is.  I would say something like:
```
Given an array `s` of ASCII codes for a sequence of hexadecimal digits, returns a `Vector{UInt8}` of bytes corresponding to the binary representation: each successive pair of hexadecimal digits in `s` gives the
value of one byte in the return vector.   (The length of `s` must be even, and the returned array has
half of the length of `s`.)
```

sambitdash replied:
change incorporated.

stevengj commented on base/strings/util.jl:
```diff
@@ -464,11 +464,89 @@ function hex2bytes(s::AbstractString)
+"""
+    hex2bytes!(d::AbstractVector{UInt8}, s::AbstractVector{UInt8})
+
+Converts the hexadecimal bytes vector to its binary representation. The results are
```
See above on the lack of clarity.  It also doesn't say which parameter is the input and which is the output.
Maybe something like:
```
Convert an array `s` of bytes representing a hexadecimal string to its binary representation, similar
to [`hex2bytes`](@ref) except that the output is written in-place in `d`.   The length of `s` must
be exactly twice the length of `d`.
```

sambitdash replied:
change incorporated.

stevengj commented on base/strings/util.jl:
```diff
@@ -464,11 +464,89 @@ function hex2bytes(s::AbstractString)
+ 0x45
+ 0x46
+
+julia> d = zeros(UInt8, 3)
```
Allocating an array via `zeros` is an unnecessary habit, and is misleading here because the reader might think it is required by `hex2bytes!`.  I would just do `hex2bytes!(Vector{UInt8}(3), s)` here.

Actually, I would prefer to see the example moved to `hex2bytes`, so you can skip the complication of allocating the input in the example.

sambitdash replied:
Allocating with zeros has been at multiple of places in examples and used in almost all test cases. It's a personal preference and style generally considered safer as a pre-initialized data.

`hex2bytes()` is an after thought which should be least used in real life. The intent to use `hex2bytes!()` is purely for IO readers who pre-allocate an array. Hence, the sample would make more sense to be there in the code which is most likely be used. Moreover, the practice of computing inside parameters is not considered ideal in examples or sample code as such. 

Hence, ignoring this comment.

sambitdash committed:
Review comments incorporated related to comments
and test cases.

sambitdash commented:
Hi All,

Before reviewing this code further I will like the reviewers to go through the following:

1. Issue #23267 
2. Discourse thread: https://discourse.julialang.org/t/need-for-a-method-hex2bytes-vector-uint8/5209 
3. Issue #14418

The intent of the introduction of this code was for IO readers to be able to read data from an text channel and pass it through a filter to a pre-allocated array. The code is written keeping in mind the performance should not be deteriorated beyond that of `hex2bytes(s::AbstractString)`. 

While conducting the exercise a very formal optimization process was not used like SIMD or acting on  multiple data the same time etc. As part of #14418, when the filters are implemented a more thorough evaluation may be carried out for these cases. Not sure if all the optimizations need to be carried out by hand or some of them may be identified by the code generation modules automatically. 

Performance is critical here as these are generally low level methods. A similar exercise was carried out while implementing hex2bytes(s::AbstractString) years' back which I modified to use here so that the code does not deteriorate on performance. Hence, if you want to change any part of the code in another way, please provide a benchmark review on the same.  

regards,

Sambit

sambitdash commented:
Here is a benchmark output on the code after removing types and additional variables:

```
julia> f_hsB_SubArray_noview8()
BenchmarkTools.Trial: 
  memory estimate:  79.99 MiB
  allocs estimate:  5242369
  --------------
  minimum time:     318.459 ms (0.85% GC)
  median time:      320.572 ms (1.12% GC)
  mean time:        321.073 ms (1.04% GC)
  maximum time:     328.791 ms (0.79% GC)
  --------------
  samples:          16
  evals/sample:     1

```
You can see the number of additional temp variable allocation just due to data conversions for a 10mb size input

Code shown below:
```
@inline number_from_hex8(c) = begin
    DIGIT_ZERO     = UInt8('0')
    DIGIT_NINE     = UInt8('9')
    LATIN_UPPER_A  = UInt8('A')
    LATIN_UPPER_F  = UInt8('F')
    LATIN_A        = UInt8('a')
    LATIN_F        = UInt8('f')

    return (DIGIT_ZERO <= c <= DIGIT_NINE) ? c - DIGIT_ZERO :
        (LATIN_UPPER_A <= c <= LATIN_UPPER_F) ? c - LATIN_UPPER_A + 10 :
        (LATIN_A <= c <= LATIN_F) ? c - LATIN_A + 10 :
        throw(ArgumentError("Not a hexadecimal number"))
end

function hex2bytes8!(d::AbstractVector{UInt8}, s::AbstractVector{UInt8})
    if 2length(d) != length(s)
        isodd(length(s)) && throw(ArgumentError("input hex array must have even length"))
        throw(ArgumentError("output array must be half length of input array"))
    end
    i, j = start(s), 0
    while !done(s, i)
        c1, i = next(s, i)
        c2, i = next(s, i)
        d[j+=1] = number_from_hex8(c1) << 4 + number_from_hex8(c2)
    end
    return d
end

f_hsB_SubArray_noview8() = begin
    const mb_10 = (10 << 20)

    arr=UInt8[rand(['0','1','2','3','4','5','6','7','8','9','0','a','b','c','d','e','f'])
                for x = 1:mb_10]
    arr1=zeros(UInt8, (10<<19))

    @benchmark hex2bytes8!(arr1,arr)
end

```

sambitdash commented:
Here is the benchmark of code which was sanitized for input and not errors rejected and implemented with word boundary computation using reinterpret. 
```
julia> f_hsB_Array_noview64()
BenchmarkTools.Trial: 
  memory estimate:  160 bytes
  allocs estimate:  4
  --------------
  minimum time:     12.517 ms (0.00% GC)
  median time:      12.641 ms (0.00% GC)
  mean time:        12.675 ms (0.00% GC)
  maximum time:     14.972 ms (0.00% GC)
  --------------
  samples:          395
  evals/sample:     1
```
Source code provided for reference:

```
@inline sanitized_number_from_hex64(c::UInt64) = begin
    n::UInt64 = (c & 0x4F4F4F4F4F4F4F4F)

    f1::UInt64 = 0x0000000000000040
    f2::UInt64 = 0xffffffffffffff07
    f3::UInt64 = 0xffffffffffffff07
    ff::UInt64 = 0xffffffffffffffff
    v::UInt64  = 0x0000000000000009

    for i = 1:8
        if (n & f1 == f1)
            n = (n & f3) + v
        end
        f1 <<= 8
        f2 <<= 8
        f3 = f2 + (ff >> 8(8-i))
        v  <<= 8
    end
    d::UInt = 0
    c1::UInt = 0
    c2::UInt = 0
    for i = 1:4
        f1 = 0xff
        c1 = (n & 0xff)
        n >>= 8
        c2 = (n & 0xff)
        n >>= 8
        d += ((c1 << 4 + c2)&0xff) << 8(i-1)
    end
    return UInt32(d)
end

function hex2bytes64!(d::Vector{UInt8}, s::Vector{UInt8})
    r = rem(length(s), 8)
    si = reinterpret(UInt64, s)
    len = div(length(s), 8)
    di = reinterpret(UInt32, d)
    @simd for i = 1:len
        di[i] = sanitized_number_from_hex64(si[i])
    end
    r_2 = div(r,2)
    i = 8len
    for j = 1:r_2
        @inbounds c1 = UInt(s[i+=1])
        @inbounds c2 = UInt(s[i+=1])
        n = sanitized_number_from_hex(c1)
        n <<= 4
        n += sanitized_number_from_hex(c2)
        @inbounds d[8len+j] = (n & 0xFF)
    end
    return d
end

@inline sanitized_number_from_hex(c::UInt) = begin
    DIGIT_NINE     = UInt('9')
    return (c > DIGIT_NINE) ? ((c & 0x07) + 9) : (c & 0x0F)
end

f_hsB_Array_noview64() = begin
    const mb_10 = (10 << 20)

    arr=UInt8[rand(['0','1','2','3','4','5','6','7','8','9','0','a','b','c','d','e','f']) for x = 1:mb_10]
    arr1=zeros(UInt8, (10<<19))

    @benchmark hex2bytes64!(arr1,arr)
end
```

stevengj commented:
Is there a type instability?  What does `@code_warntype` say?

For example, I notice that the `+ 10` is a problem because that incurs a promotion to `UInt`.  Add `0x0a` instead.

stevengj commented:
(Adding a bunch of explicit type declarations in Julia is usually a sign that you are making a mistake somewhere.   Often it is masking an unintended type conversion somewhere.   And it doesn't make sense to do all computations with 64-bit quantities when all the actual values are 8-bit; probably the performance problem is because of your unintended conversions to 64-bit via your use of `Int` literals.)

sambitdash commented:
@stevengj There is no performance problem with computations with UInt64 quantities but with usage of UInt8 functions as suggested. Secondly, is it documented that 0x0a will be interpreted differently than 10, although typeof on REPL provides UInt8. Thirdly, is there a benchmark to suggest the changes will be actually improving on performance. If not I will be reluctant to make any further changes to the code.

stevengj commented:
I verified on my machine that changing `+ 10` to `+ 0x0a` in the `UInt8` version eliminates the type instability (`@code_warntype` warns about it!) and makes the routine almost 10x faster.

With some additional cleanups, my code is:
```jl
@inline number_from_hex8(c) =
    (UInt8('0') <= c <= UInt8('9')) ? c - UInt8('0') :
    (UInt8('A') <= c <= UInt8('F')) ? c - (UInt8('A') - 0x0a) :
    (UInt8('a') <= c <= UInt8('f')) ? c - (UInt8('a') - 0x0a) :
    throw(ArgumentError("byte is not a ASCII hexadecimal digit"))

function hex2bytes8!(d::AbstractVector{UInt8}, s::AbstractVector{UInt8})
    if 2length(d) != length(s)
        isodd(length(s)) && throw(ArgumentError("input hex array must have even length"))
        throw(ArgumentError("output array must be half length of input array"))
    end
    i, j = start(s), start(d)
    while !done(s, i)
        @inbounds d[j] = number_from_hex8(s[i]) << 4 + number_from_hex8(s[i+1])
        i += 2
        j += 1
    end
    return d
end

```

stevengj commented:
@sambitdash, the performance problem is not with `UInt8`, but rather with the type instability; see my improved version above and benchmarks thereof.  Regarding integer literals, see the discussion [in the manual](https://docs.julialang.org/en/latest/manual/integers-and-floating-point-numbers/#Integers-1): the literal `10` has type `Int` and `0x0a` has type `UInt8`, and the result type of `+` is determined by [promoting the operands](https://docs.julialang.org/en/latest/manual/conversion-and-promotion).

sambitdash commented:
@stevengj thanks. As a maintainer you have the rights to the make edits. Hence, please submit the changes as desired.

stevengj commented:
(Of course, you can make it faster still by assuming valid data, etcetera, but I don't think the additional gains are worth it.)

stevengj commented:
@sambitdash, I don't think I can push changes to this PR, because the PR is pulled from [your fork](https://github.com/sambitdash/julia/tree/I23161).

fredrikekre commented:
Seems like the "Allow edits from maintainers." checkbox is ticked though, at least I can make edits :)

sambitdash committed:
Checking in code provided by stevengj

There are some test case failures reported related to views.

sambitdash commented:
@stevengj  and @fredrikekre , I checked in the code @stevengj  provided earlier. Those conditions break the test cases on views due to linear indexing assumptions. 

@fredrikekre, please make changes necessary related to the code and test cases as desired.

stevengj committed:
Update util.jl

further condense implementation and documentation, merging with `hex2bytes(s::AbstractString)`

stevengj commented on base/strings/util.jl:
```diff
@@ -499,29 +453,38 @@ julia> s = b"01abEF"
 """
+function hex2bytes end
+
+hex2bytes(s::AbstractString) = hex2bytes(Vector{UInt8}(String(s)))
```
This makes an additional copy of the data, but is still faster than the old implementation on my machine because it operates on bytes rather than unicode characters.  More importantly, it avoids code duplication.

(In any case, it's not clear to me how performance-critical this function is.)

stevengj commented:
@fredrikekre, ah, thanks.

stevengj committed:
test updates

use ===, not ==, to test that operation occurs in-place, and `hex2bytes!` now throws an error for incorrect output length

stevengj commented:
(Note that 1d views still use `start:end` indexing.   The main thing to be careful of is that you don't want to assume that indices start at 1, which may not be true for custom array types like OffsetArrays.)

stevengj commented:
(It's annoying to make more than 1-line changes via the github web UI since I have to wait for CI to run in order to detect any typos.)

stevengj committed:
fix hex2bytes for views (where start is not an index)

stevengj committed:
rm redundant length check

stevengj commented:
Travis failures seem unrelated (a timeout and a problem with `test/file.jl`)

StefanKarpinski commented:
This ended up being a very nice (first?) PR to Julia – thanks for bearing with it, @sambitdash. One simple API item that seems to be missing still is a `hex2bytes!` method that takes strings as the second argument. However, that can be added at any point should someone need it, so I'm not concerned. Thanks for the excellent work!

sambitdash commented:
Thanks @stevengj for all the help. Without your assistance I will would not have understood the Julia type system and the associated code generation and would have attempted type conversions / casting to optimize. Thanks everyone for your help in accessing `base` and understand the style guidelines.
